### PR TITLE
Add additional datasets with preprocess and testing for prostate example

### PR DESCRIPTION
**Pull request by ZiyueXu77**
Switch from 3D to 2D
Add two more datasets
Add preprocessing scripts
Add standalone testing scripts
Update Readme

### CODE REVIEW

**Comment by holgerroth** [comment]
Let's upgrade this to use the 2.1 configure style first.

**Comment by ZiyueXu77** [comment]
Added both 2D and 3D examples with updated readmes @holgerroth @Can-Zhao

**Comment by ZiyueXu77** [comment]
Lots of files but many are duplicated configs

File: examples/prostate/prostate_2D/start_fl_poc.sh [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+echo "There are ${n_gpus} GPUs."
+# start server
+echo "STARTING SERVER"
+export CUDA_VISIBLE_DEVICES=0
+bash "${workspace}/server/startup/start.sh" ${servername} &
+sleep 10
+
+# start clients
+echo "STARTING CLIENTS"
+i=1
```
Comment (holgerroth) [inline examples/prostate/prostate_2D/start_fl_poc.sh:37]: why is this starting from 1?

File: examples/prostate/prostate_3D/start_fl_poc.sh [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+echo "There are ${n_gpus} GPUs."
+# start server
+echo "STARTING SERVER"
+export CUDA_VISIBLE_DEVICES=0
+bash "${workspace}/server/startup/start.sh" ${servername} &
+sleep 10
+
+# start clients
+echo "STARTING CLIENTS"
+i=1
```
Comment (holgerroth) [inline examples/prostate/prostate_3D/start_fl_poc.sh:35]: why is this starting from 1?

File: examples/prostate/prostate_2D/README.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+The testing score is computed based on the best global model for Central/FedAvg/FedProx, and the six best personalized models for Ditto.
+We provide a script for performing validation on testing data split, please add the correct paths and job_ids, and run
+
+```
+bash ./result_stat/testing_models_2d.sh
+```
+
+The Dice results for the above run are:
+
+Centralized: 0.8590
```
Comment (holgerroth) [inline examples/prostate/prostate_2D/README.md:134]: make this a table?

File: examples/prostate/README.md [hunk 4]
```diff
@@ -117,131 +127,109 @@ Note that there is one case (ProstateDx-01-0055.nii.gz) with image/label mismatc
...
+Generate data split for ./dataset/NCI_ISBI_Dx, with train:validation:test 0.5:0.25:0.25
+Save json to ./datalist/client_NCI_ISBI_Dx.json
+Mode: file
+In total 39 cases, 20 for training, 10 for validation, and 9 for testing
+```
 
-The TensorBoard curves for validation Dice of the global model for the 1000 epochs (100 rounds, 10 local epochs per round) during training are shown below:
-![All training curve](./figs/all_training.png)
+## 3. Federated Training for Prostate Tasks
+Please go to subfolders `prostate_2D` and `prostate_3D` for further instructions on federated training.
```
Comment (holgerroth) [inline examples/prostate/README.md:235]: make these hyperlinks?

File: examples/prostate/prostate_3D/README.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+Let's summarize the result of the experiments run above. We compare the validation scores of 
+the global model for Central/FedAvg/FedProx, and personalized models for Ditto. In this example, each client computes their validation scores using their own
+validation set, and the centralized model computes the validation score using the combined validation set. Please note that due to the limited size of data set, the validation curve can have significant variations across runs.
+
+We provide a script for plotting the tensorboard records, running
+```
+python3 ./result_stat/plot_tensorboard_events.py
+```
+The TensorBoard curves for validation Dice for the 150 epochs (150 rounds, 1 local epochs per round) during training are shown below:
+![All training curve](./figs/all_training.png)
```
Comment (holgerroth) [inline examples/prostate/prostate_3D/README.md:122]: could add a table with results here as well as in the 2D example.

**Review by holgerroth — CHANGES_REQUESTED** [review]

File: examples/prostate/prostate_2D/configs/prostate_ditto/config/config_train.json [hunk 6]
```diff
@@ -0,0 +1,9 @@
+{
+  "learning_rate": 1e-3,
+  "ditto_learning_rate": 1e-3,
+  "ditto_lambda": 1.0,
+  "fedproxloss_mu": 0.0,
+  "cache_dataset": 1.0,
+  "dataset_base_dir": "PWD/../data_preparation/dataset_2D",
```
Comment (holgerroth) [inline examples/prostate/prostate_2D/configs/prostate_ditto/config/config_train.json:7]: use {datafolder} alias instead of hardcoded path.

File: examples/prostate/prostate_2D/start_fl_poc.sh (hunk 1 above)
Comment (ZiyueXu77) [reply examples/prostate/prostate_2D/start_fl_poc.sh]: changed to 0

File: examples/prostate/prostate_3D/start_fl_poc.sh (hunk 2 above)
Comment (ZiyueXu77) [reply examples/prostate/prostate_3D/start_fl_poc.sh]: changed to 0

File: examples/prostate/README.md (hunk 4 above)
Comment (ZiyueXu77) [reply examples/prostate/README.md:235]: sounds good, changed

File: examples/prostate/prostate_2D/configs/prostate_ditto/config/config_train.json (hunk 6 above)
Comment (ZiyueXu77) [reply examples/prostate/prostate_2D/configs/prostate_ditto/config/config_train.json]: changed

File: examples/prostate/prostate_2D/README.md (hunk 3 above)
Comment (ZiyueXu77) [reply examples/prostate/prostate_2D/README.md]: changed

File: examples/prostate/README.md [hunk 7]
```diff
@@ -44,17 +41,27 @@ cd ./data_preparation
...
   mkdir Raw/${dataset}
 done
 ```
 
-### 2.1 Downloading steps for each dataset: 
+Note that for 2D example, we use all datasets, while for 3D example we use 4 of them to limit the GPU resource cost. 
+
+Various packages are needed for the candidate datasets, we will need the following:
+ ```
+pip3 install dicom2nifti
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/prostate/README.md:61]: added in another PR, in case someone needs to review it

File: examples/prostate/prostate_3D/README.md (hunk 5 above)
Comment (ZiyueXu77) [reply examples/prostate/prostate_3D/README.md:123]: agree, will add

Comment (ZiyueXu77) [reply examples/prostate/prostate_3D/README.md:124]: added, results different from last version because last time the split is 0.8:0.2:0.0 and current split is 0.5:0.25:0.25

**Review by holgerroth — CHANGES_REQUESTED** [review]

File: examples/prostate/prostate_3D/README.md [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+In this example, we run four clients on two GPUs, two clients for each GPU with 12 GB memory.
+
+Then FL training will be run with an automatic script utilizing the FLAdminAPI functionality.    
+The [submit_job.sh](./submit_job.sh) script follows the pattern:
+```
+bash ./submit_job.sh [config]
+```
+`[config]` is the experiment job that will be submitted for the FL training, in this example, this includes `prostate_central`, `prostate_fedavg`, `prostate_fedprox`, and `prostate_ditto`.  
+
+Note that in order to make it working under most system resource conditions, the current script used regular `Dataset` for data loading in `pt/learners/prostate_learner.py`, which could be slow. If resource permits, it will make the training much faster by replacing it with `CacheDataset`. More information available [here](https://docs.monai.io/en/stable/data.html#cachedataset).
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/prostate/prostate_3D/README.md:81]: yes, changed

File: examples/prostate/virtualenv/plot-requirements.txt [hunk 9]
```diff
@@ -0,0 +1,2 @@
+tensorflow
+seaborn
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/prostate/virtualenv/plot-requirements.txt:2]: yes, added

File: examples/prostate/result_stat/prostate_3d_test_only.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
+# Copyright 2020 MONAI Consortium
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/prostate/result_stat/prostate_3d_test_only.py:1]: right I think nvflare makes sense since it is under nvflare repo, updated other license as well

**Review by holgerroth — APPROVED** [review]
Nice job!
